### PR TITLE
[WIP] swconfig-lua

### PR DESCRIPTION
**Pull request by mwetterw**
Note: We'll need to **change the base branch** of the merge above. It's currently v0.4 and should be changed.

### CODE REVIEW

File: swconfig/files/switch.lua [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+
+  self.sock = sock
+  return true
+end
+
+-- This closes the serial connection and resets the switch object for reuse
+-- It's very important NOT to logout from the switch
+-- If we logout 5 times, we get following warning:
+--  CLI is restarting too fast
+--  Console will be blocked 300 seconsd if you restart CLI less than 90 seconds next time.
```
Comment (gregdel) [inline swconfig/files/switch.lua]: typo

File: swconfig/files/switch.lua [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+--
+-- If we logout a sixth time, UART CLI won't respond at all for 5 minutes
+--  Console is blocking 300 seconsd
```
Comment (gregdel) [inline swconfig/files/switch.lua]: same typo :P

File: swconfig/files/switch.lua [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+    self.sock = nil
+  end
+
+  self.hostname = nil
+  self.state = Switch.State.UNKNOWN
+end
+
+-- Ask the kernel to see if someone else is already using the serial link
+function Switch:_file_is_busy()
+  fd = io.popen("lsof " .. self.file)
```
Comment (gregdel) [inline swconfig/files/switch.lua]: Don't you need to check the value of fd ? I'm no lua expert but I can see people do things like `assert(io.popen("your_program"))`

File: swconfig/files/switch.lua [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+  return data
+end
+
+-- This low level method sends a command to the switch and consumes the echo
+function Switch:_send(cmd)
+  -- When sending a command, only append a Line Feed
+  cmd = cmd .. "\n"
+
+  for i = 1, #cmd do
+    char = string.sub(cmd, i, i)
```
Comment (gregdel) [inline swconfig/files/switch.lua]: LOL cmd[i] does not exist in lua

File: swconfig/files/switch.lua [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+  end
+
+  -- TODO: Handle the --More--
+
+  return res
+end
+
+-- This method analyzes the received output to determine the switch state
+function Switch:_parse_prompt(data)
+  first_line = data[1]
```
Comment (gregdel) [inline swconfig/files/switch.lua]: A table starts at 1 ? lua ? Mouahaha

File: swconfig/files/switch.lua [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+    print("State 'Press any key' detected")
+    return Switch.State.PRESS_ANY_KEY
+  end
+
+  -- Determine the hostname
+  if self.hostname == nil then
+    -- '%' is the lua escape character for pattern, because '(' otherwise would have another meaning
+    needles = {"%(", "# ", "> "}
+    for key, needle in ipairs(needles) do
+      hostname_end = string.find(last_line, needle)
```
Comment (gregdel) [inline swconfig/files/switch.lua]: Maybe we should use "plain" to disable the pattern matching feature

Comment (gregdel) [inline swconfig/files/switch.lua]: `first_char, _  := string.find(.....)` to be clear that you ignore the last return

File: swconfig/files/switch.lua (hunk 2 above)
Comment (mwetterw) [inline swconfig/files/switch.lua]: I'm quoting what the switch says. The switch makes typos, so I quote typos =)

**Comment by mwetterw** [comment]
Bye Bye Dear PR :-)

(See #329 )

**Pull request closed without merging**